framework/src/media/MediaRecorderImpl.cpp

```diff
@@ -475,23 +483,83 @@ void MediaRecorderImpl::setRecorderObserver(std::shared_ptr<MediaRecorderObserve
 	notifySync();
 }
 
+recorder_result_t MediaRecorderImpl::setDuration(int second)
+{
+	std::unique_lock<std::mutex> lock(mCmdMtx);
+
+	medvdbg("MediaRecorderImpl::setDuration()\n");
+	RecorderWorker& mrw = RecorderWorker::getWorker();
+	if (!mrw.isAlive()) {
+		meddbg("Worker is not alive\n");
+		return RECORDER_ERROR;
+	}
+
+	recorder_result_t ret = RECORDER_ERROR;
+	mrw.enQueue(&MediaRecorderImpl::setRecorderDuration, shared_from_this(), second, std::ref(ret));
+	mSyncCv.wait(lock);
+
+	return ret;
+}
+
+void MediaRecorderImpl::setRecorderDuration(int second, recorder_result_t& ret)
+{
+	medvdbg("setRecorderDuration mCurState : %d\n", (recorder_state_t)mCurState);
+
+	if (mCurState != RECORDER_STATE_IDLE) {
+		meddbg("setRecorderDuration Failed mCurState: %d\n", (recorder_state_t)mCurState);
+		return notifySync();
+	}
+	if (second > 0) {
+		medvdbg("second is greater than zero, set limit : %d\n", second);
+		mDuration = second;
+	}
+
+	ret = RECORDER_OK;
+	notifySync();
+}
+
 void MediaRecorderImpl::capture()
 {
 	medvdbg("MediaRecorderImpl::capture()\n");
-	int frames = start_audio_stream_in(mBuffer, get_input_frame_count());
+	unsigned int frameSize = get_input_frame_count();
 
+	if (mTotalFrames > 0) {
+		int64_t remainFrames = mTotalFrames - mCapturedFrames;
+		/* almost reaches size limit, adjust frame size before request capture to audio_manager */
+		if (remainFrames < frameSize) {
+			frameSize = remainFrames;
+		}
+	}
+	
+	int frames = start_audio_stream_in(mBuffer, frameSize);
 	if (frames > 0) {
-		int size = get_input_frames_to_byte(frames);
-
+		mCapturedFrames += frames;
+		if (mCapturedFrames > INT_MAX) {
+			meddbg("Too huge value : %d, set 0 to prevent overflow\n", mCapturedFrames);
```

Comment:
Error log is displayed to set the mCapturedFrames to 0
But it's not been done in the code, who has to set it to 0 and when ?

Reply:
Thanks. I missed it.